### DISCUSSION TITLE
2.4.0 - Flakey Connection to the Ryuk

### DISCUSSION THREAD

**Discussion by cdconn00** [testcontainers/testcontainers-dotnet#775] [Q&A] [answered]
Howdy, @HofmeisterAn! Experiencing a similar issue to #443 but in version 2.4.0. I didn't see a solution other than some saying a version upgrade resolved the issue. Being a few versions ahead now I'm still running into the same issue. In other projects I have circumvented this error by elevating Rancher/Docker Desktop to Admin, but that is unfortunately not an option in my current project. 

~ every 3 runs test containers is functioning correctly, but the other times it fails to connect to ryuk and the container shows (like in #443):
![image](https://user-images.githubusercontent.com/30357632/218114395-28d63c53-978d-4adb-b976-9a6da0ae7a79.png)

Below is an implementation with some things I have tried uncommented.
```
using Microsoft.EntityFrameworkCore;
using System;
using Project.Persistence;
using DotNet.Testcontainers.Builders;
using System.Threading.Tasks;
using Xunit;
using DotNet.Testcontainers.Configurations;
using DotNet.Testcontainers.Containers;
using System.Net.Sockets;
using System.Net;

namespace Project.Application.Tests
{
    public class TestBase : IAsyncLifetime
    {
        private ProjectDbContext _dbContext;
        public const string TemplateDatabaseName = "TestDB";

        private readonly PostgreSqlTestcontainer _dbContainer = new ContainerBuilder<PostgreSqlTestcontainer>()
            .WithDatabase(new PostgreSqlTestcontainerConfiguration
            {
                Database = "db",
                Username = "postgres",
                Password = "postgres",
            })
            .WithImage("postgres:15.1-alpine")
            //.WithPortBinding(5432, true)
            //.WithWaitStrategy(Wait.ForUnixContainer()
            //    .UntilPortIsAvailable(5432))

            //.WithEnvironment("REAPER_PORT", FreeTcpPort().ToString()) // Realized this was for a different project... possible in in testcontainers-dotnet?

            .Build();

        public virtual async Task InitializeAsync()
        {
            await GetDbContextAsync();
        }

        private async Task GetDbContextAsync()
        {
            // TestcontainersSettings.ResourceReaperEnabled = false; // no difference in runs, get the same resourcereaperexception
            await _dbContainer.StartAsync();

            var options = new DbContextOptionsBuilder<ProjectDbContext>()
                        .UseNpgsql(_dbContainer.ConnectionString, n => n.MigrationsAssembly("Project.Persistence"))
                        .Options;
            AppContext.SetSwitch("Npgsql.EnableLegacyTimestampBehavior", true);
            _dbContext = new ProjectDbContext(options);
            _dbContext.Database.EnsureDeleted();
            _dbContext.Database.EnsureCreated();
            _dbContext.Database.CloseConnection();

            await _dbContext.Database.MigrateAsync();
        }

        public async Task DisposeAsync()
        {
            await Context.Database.EnsureDeletedAsync();
        }

        public ProjectDbContext Context => _dbContext;

        static int FreeTcpPort()
        {
            TcpListener l = new TcpListener(IPAddress.Loopback, 0);
            l.Start();
            int port = ((IPEndPoint)l.LocalEndpoint).Port;
            l.Stop();
            return port;
        }
    }
}

```
From what I've read it looks like specifying a port for the Ryuk container could be a solution, is there support for that in 2.4.0?

Apologies if there's a solution I'm overlooking, I've been through several discussions/issues but the solutions have not seemed to fix the issues. Thank y'all for maintaining this project it has been an absolute delight on several projects I've worked with.

Relevant Versioning/Info:
- Unable to run VisualStudio,DockerDesktop,etc at an elevated level
- TestContainers 2.4.0
- Windows 11 Enterprise
- Docker Desktop 4.16.3 (96739) running WSL2 engine

**Comment by HofmeisterAn** [comment, marked as answer]
> ~ every 3 runs test containers is functioning correctly, but the other times it fails to connect to ryuk and the container shows (like in #443):

That is way too frequent. Sounds like another issue. I have seen some issues with Hyper-V or the machine network configurations. Did you check if you run into the excluded port ranges: https://github.com/docker/for-win/issues/3171.

Your configuration looks a bit unfortunate. I suggest to run a simple test to figure out what is causing the issue on your machine.

```csharp
public class GitHub : IAsyncLifetime
{
    private readonly IContainer _container = new ContainerBuilder()
        .WithImage("alpine:3.17")
        .WithEntrypoint("/bin/sh", "-c")
        .WithCommand("trap : TERM INT; sleep infinity & wait")
        .Build();

    public Task InitializeAsync()
    {
        return _container.StartAsync();
    }

    public Task DisposeAsync()
    {
        return _container.DisposeAsync().AsTask();
    }

    [Fact]
    public Task Discussions775()
    {
        return Task.CompletedTask;
    }
}
```

Can you add Testcontainers log output including the stack trace? Did you check which port Ryuk is using and Testcontainers try to connect to?

**Reply by cdconn00** [reply]
Thanks for the quick reply sir.

Looks like I have run into the excluded port ranges, thanks for the reference:
![image](https://user-images.githubusercontent.com/30357632/218191139-764423b5-cd87-4ccd-9483-a0ec4212d0e9.png)

ST:
```
Message:
DotNet.Testcontainers.Containers.ResourceReaperException : Initialization has been cancelled.

Stack Trace:
ResourceReaper.GetAndStartNewAsync(Guid sessionId, IDockerEndpointAuthenticationConfiguration dockerEndpointAuthConfig, IImage resourceReaperImage, IMount dockerSocket, Boolean requiresPrivilegedMode, TimeSpan initTimeout, CancellationToken ct)
ResourceReaper.GetAndStartNewAsync(Guid sessionId, IDockerEndpointAuthenticationConfiguration dockerEndpointAuthConfig, IImage resourceReaperImage, IMount dockerSocket, Boolean requiresPrivilegedMode, TimeSpan initTimeout, CancellationToken ct)
ResourceReaper.GetAndStartDefaultAsync(IDockerEndpointAuthenticationConfiguration dockerEndpointAuthConfig, CancellationToken ct)
TestcontainersClient.RunAsync(IContainerConfiguration configuration, CancellationToken ct)
DockerContainer.UnsafeCreateAsync(CancellationToken ct)
DockerContainer.UnsafeStartAsync(CancellationToken ct)
TestContainersDbug.InitializeAsync() line 32
```

I wasn't able to get the TC logs from redirecting stdout/err with Consume or my IDEs debug output, apologies.

Regardless it seems like I am running into the excluded ports. I'll likely be unable to proceed, as using the workarounds won't be possible for me in this case. Thanks again for looking into this.

**Reply by HofmeisterAn** [reply]
The referenced issue contains a few suggestions how to deal with the excluded port ranges, none of them help?

----

You can override the Resource Reaper port resolution, but this is **not recommended**. The example below is not recommended. You should fix your machine's configuration, you will run into other issues otherwise.

```csharp
public class GitHub : IAsyncLifetime
{
    static GitHub()
    {
        TestcontainersSettings.ResourceReaperPublicHostPort = _ =>
        {
            using (var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
            {
                socket.Bind(new IPEndPoint(IPAddress.Loopback, 0));
                return (ushort)((IPEndPoint)socket.LocalEndPoint).Port;
            }
        };
    }

    private readonly IContainer _container = new ContainerBuilder()
        .WithImage("alpine:3.17")
        .WithEntrypoint("/bin/sh", "-c")
        .WithCommand("trap : TERM INT; sleep infinity & wait")
        .Build();

    public Task InitializeAsync()
    {
        return _container.StartAsync();
    }

    public Task DisposeAsync()
    {
        return _container.DisposeAsync().AsTask();
    }

    [Fact]
    public Task Discussions775()
    {
        return Task.CompletedTask;
    }
}
```

**Reply by cdconn00** [reply]
Thank you, sir, I appreciate you! Those solutions will work -- it's more of a team/permissions issue for not pursuing them on the current project.

Still loving test containers in my other projects -- thanks for all the work you do!
